docs/hll/hll.tex

```diff
@@ -0,0 +1,287 @@
+massive streams (read-once sequences). This problem has received a great deal of
+attention over the past two decades, finding an ever growing number of
+applications in networking and traffic monitoring, such as the detection of worm
+propagation, of network attacks (e.g., by Denial of Service), and of link-based
+spam on the web.}
+
+\subsection{Why HyperLogLog?}
+
+So, what's a good use case for HLLs? One example would be to determine the
+number of distinct search queries on \textit{google.com} over a time
+period\cite{Heule:2013:Online}.
+
+\indent The goal of HLL is to estimate unique elements in large sets (large being beyond
+$10^{9}$) and streams while also keeping memory low(er). Normally, calculating
+the exact cardinality of a set requires an amount of memory proportional to the
+cardinality when counting these unique items. With HLLs, the trade off is less
+memory in exchange for approximated cardinality. Yo performance.\newline
+
+As per \cite{Heule:2013:Online}, the key requirements for cardinality an
+estimation algorithm are
+
+\begin{enumerate}
+\item \textbf{Accuracy}: For a fixed amount of memory, the algorithm should
+  provide as accurate an estimate as possible. Especially for small
+  cardinalities, the results should be near exact.
+\item \textbf{Memory efficiency}: The algorithm should use the available memory
+  efficiently and adapt its memory usage to the cardinality. That is, the
+  algorithm should use less than the user-specified maximum amount of memory if
+  the cardinality to be estimated is very small.
+\item \textbf{Estimate large cardinalities}: Multisets with cardinalities well
+  beyond 1 billion occur on a daily basis, and it is important that such large
+  cardinalities can be estimated with reasonable accuracy.
+\item \textbf{Practicality}: The algorithm should be implementable and
+  maintainable.
+\end{enumerate}
+
+There are two generalized categories of cardinality observables
+\cite{Flatjolet:2007:Online}:
+
+\begin{enumerate}
+\item \textbf{Bit-pattern observables}: these are based on certain patterns of
+  bits occurring at the beginning of the (binary) $S$-values. For instance,
+  observing in the stream $S$ at the beginning of a string a bitpattern
+  $0^{p-1}1$ is more or less a likely indication that the cardinality $n$
+  of $S$ is at least $2^{p}$. HLL is an example of this category.
+
+\item \textbf{Order statistics observables}: these are based on order statistics,
+  like the smallest (real) values, that appear in $S$. For instance,
+  if $X = min(S)$, we may legitimately hope that $n$ is roughly of the order of
+  $1/X$, since, as regards expectations, one has $\mathbb{E}(X) = 1/(n + 1)$.
+\end{enumerate}
+
+\section{The HyperLogLog Algorithm}
+
+\indent The key components of the algorithm are
+
+\begin{enumerate}
+\item \textit{randomization achieved by a hash function}, $h$, which is applied
+  to every element which is counted bits.
```

Comment:
hard for me to understand what is meant by "which is applied to every element which is counted bits"